program/src/processor.rs

```diff
@@ -847,13 +970,24 @@ impl Processor {
             return Err(ProgramError::InvalidArgument);
         }
 
+        let effective_interest =
+            obligation.calc_effective_interest_amount(clock.slot, liquidity.interest)?;
+        msg!("effective_interest: {}", effective_interest);
+
+        // 0. First we repay effective interest
+        obligation.update_interest_amount(effective_interest.checked_sub(amount).unwrap_or(0));
+        obligation.update_slot(clock.slot);
+
+        // 1. And then we repay borrowed liquidity
+        let repay_amount = amount.checked_sub(effective_interest).unwrap_or(0);
         let repay_limit = obligation.amount_liquidity_borrowed;
-        if amount > repay_limit {
+        if repay_amount > repay_limit {
```

Comment:
Maybe in this case it is better to charge user the maximum amount allowed instead of throwing exception? If you agree, it makes sense to rename amount into max_amount.